### ISSUE TITLE
Clang sometimes fails to complete after 4a34571b52e088d2d6d8d362b76586d9cca466a8

### ISSUE THREAD

**Issue by Havner** [company-mode/company-mode#749]
I've attached a very simple file where clangs fails to complete. Try to complete at the end of `yaca_fr` line. It fails only after https://github.com/company-mode/company-mode/commit/4a34571b52e088d2d6d8d362b76586d9cca466a8. Reverting it fixes the issue completely.

Some investigation I made:

If I try to complete at `yaca_` it succeeds. But If I try to do that after `yaca_fr` it fails. In the clang-error I can see:

```
clang failed with error 1:
/usr/bin/clang -fsyntax-only -Xclang -code-completion-macros -x c -I/usr/local/include/yaca -I/usr/include/python3.5m -Xclang -code-completion-at=-:7:9 -

<stdin>:7:2: error: use of undeclared identifier 'yaca_fr'; did you mean 'yaca_free'?
        yaca_fr
        ^~~~~~~
        yaca_free
<stdin>:3:6: note: 'yaca_free' declared here
void yaca_free(void *memory);
     ^
1 error generated.
```

What's interesting though is that when it succeeds (at `yaca_`) it also returns error (I've checked this in cmd line), but doesn't try to to suggest what I meant. E.g.:

```
<stdin>:7:2: error: use of undeclared identifier 'yaca_'
        yaca_
        ^
1 error generated.
```

Then it works. The moment that clang tries to suggest what function I meant it stops giving completions. Reverting the mentioned patch fixes the issue completely. Although I don't know why.

```
$ clang --version
clang version 3.9.1 (tags/RELEASE_391/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

[comp-clang-bug.c.gz](https://github.com/company-mode/company-mode/files/1568815/comp-clang-bug.c.gz)

**Comment by dgutov** [comment]
Seems to work fine here. @Havner Any change you can try this with Clang 4.0 as well?

**Comment by Havner** [comment]
I don't have 4.0 anywhere, but I do have macOS 10.13 where clang is:

```
$ clang --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

And yes, it works there properly. So it really seems to be a bug in Clang as I was afraid.

Anyway, 2 months ago it did work in company, it stopped after this patch and I believe lots of people might be using older clang. I don't really have means to upgrade it at work. Maybe some compatibility option in company-clang so it can work on older ones as it used to before?

**Comment by dgutov** [comment]
We can do a version check. Unfortunately, it will be non-trivial for Apple Clang where they have a separate version history. Do you know if Apple LLVM 8.0 is still in wide use? Apparently it's based on Clang 3.9.

**Comment by Havner** [comment]
I think it's irrelevant what LLVM 9.0 is based upon. It probably is 4.0 if 4.0 works for you fine and it works fine here as well. What worries me more are Linux distributions where < 4.0 is still widely used. In my company we use Ubuntu 16.04 which is still the latest LTS and there is no Clang 4.0.

It would really be beneficial for a small thing like this do provide some fallback (might be explicitly configured with an option) to make this work. In a year or so it could probably be removed.

**Comment by dgutov** [comment]
Like I said, we can do a version check. It won't do for every affected user to have to configure it for themselves.

**Comment by dgutov** [comment]
I'm asking about older Apple LLVM versions. Version 8.0 and lower.

**Comment by Havner** [comment]
Sorry, I misunderstood you. You mean do a version check inside the company-clang.

**Comment by Havner** [comment]
I honestly don't know how widely used they are.

For macOS what I do have are the latest XCode command line tools which might be related to XCode version which is 9.2. I had them on macOS 10.12 as well, but no idea if the clang was 9.0 as well as here. Have no means to check that now.

I think it'd be safe to consider >= 4.0 on linux and >= llvm 9.0 on macOS to be the fixed ones.

**Comment by dgutov** [comment]
Should be fixed now. Please try it out.

**Comment by Havner** [comment]
For some reason I cannot get any completion now but I can't figure out why. There is no error for clang (no clang buffers gets created).

`(company-clang--check-version 4.0 9.0)` evaluates to `t`.
`(company-clang-version)` evaluates to `(apple . 9.0)`.

So it should be fine. I'll test this tomorrow at work and will give it one more go on macOS later.

**Comment by Havner** [comment]
Something is definitely wrong on macOS, because reverting to https://github.com/company-mode/company-mode/commit/4004c7f3ccd4f2fdede71f1b82216e93dbbf6826 makes it all work back again.

**Comment by Havner** [comment]
I think this expression is not correct now:

`(< company-clang--version company-clang-required-version)`

e.g.
`(<(apple . 9.0) 1.1)`

**Comment by dgutov** [comment]
Ouch! Should be fixed now. Please try again.

**Comment by Havner** [comment]
It works on clang 3.9 I have. I'll be testing this on 3.8 this week as well. And will let you know about the Apple one, but seeing yesterday it should work after you've fixed this one exp.

Cheers!

**Comment by Havner** [comment]
Works fine on macOS now.

**Comment by dgutov** [comment]
Thanks for checking!
